### PR TITLE
MRG: EMS transformer with sklearn API

### PR DESCRIPTION
**Pull request by kingjr**
closes https://github.com/mne-tools/mne-python/issues/3427

- [X] Add EMS transformer and test
- [X] Update example
- [X] Add recommendations in the example that diverge from the original paper.

### CODE REVIEW

File: examples/decoding/plot_ems_filtering.py [hunk 1]
```diff
@@ -57,38 +60,63 @@
...
+filters = list()  # Spatial filters at each time point
+
+# In the original paper, the cross-validation is a leave-one-out. However,
+# we recommend using a Stratified KFold, as LOO is known to generate biases.
+for train, test in StratifiedKFold(y):
+    # In the original paper, the z-scoring is applied outside the CV.
+    # However, we recommend to apply this preprocessing inside the CV.
+    # Note that such scaling should be done separately for each channels if the
+    # data contains multiple channel types.
+    X_scaled = X / np.std(X)  # X_scaled does not have any unit anymore
```
Comment (agramfort) [inline examples/decoding/plot_ems_filtering.py]: this scaling is the same for every fold

Comment (agramfort) [inline examples/decoding/plot_ems_filtering.py]: I assume you want to do it outside of the loop

Comment (agramfort) [inline examples/decoding/plot_ems_filtering.py]: or maybe you want to do X / np.std(X[train])

Comment (kingjr) [inline examples/decoding/plot_ems_filtering.py]: yes my bad, I was checking that the figures was identical to the original example, but I think we should indeed advice for `X[train] / np.std(X[train])`

File: examples/decoding/plot_ems_filtering.py [hunk 2]
```diff
@@ -57,38 +60,63 @@
 reject = dict(grad=4000e-13, eog=150e-6)
 
 epochs = mne.Epochs(raw, events, event_ids, tmin, tmax, picks=picks,
-                    baseline=None, reject=reject)
-
-# Let's equalize the trial counts in each condition
-epochs.equalize_event_counts(epochs.event_id, copy=False)
-
-# compute surrogate time series
-surrogates, filters, conditions = compute_ems(epochs, ['AudL', 'VisL'])
```
Comment (kingjr) [inline examples/decoding/plot_ems_filtering.py]: fair enough

File: examples/decoding/plot_ems_filtering.py [hunk 3]
```diff
@@ -57,38 +60,63 @@
...
 plt.figure()
 plt.title('Average EMS signal')
-
-mappings = [(k, v) for k, v in event_ids.items() if v in conditions]
+mappings = [(key, value) for key, value in event_ids.items()]
 for key, value in mappings:
-    ems_ave = surrogates[conditions == value]
-    ems_ave *= 1e13
-    plt.plot(times, ems_ave.mean(0), label=key)
+    ems_ave = X_transform[y == value]
```
Comment (kingjr) [inline examples/decoding/plot_ems_filtering.py]: note that this re-scaling isn't right. after normalization the data looses its unit. More generally, EMS can be applied across different sensor type and thus loses its unit.

File: examples/decoding/plot_ems_filtering.py [hunk 4]
```diff
@@ -57,38 +60,71 @@
...
+
+# In the original paper, the cross-validation is a leave-one-out. However,
+# we recommend using a Stratified KFold, as LOO is known to generate biases.
+for train, test in StratifiedKFold(y):
+    # In the original paper, the z-scoring is applied outside the CV.
+    # However, we recommend to apply this preprocessing inside the CV.
+    # Note that such scaling should be done separately for each channels if the
+    # data contains multiple channel types.
+    # Note that X_scaled does not have any unit anymore.
+    X_scaled = X / np.std(X[train])
```
Comment (jona-sassenhagen) [inline examples/decoding/plot_ems_filtering.py]: Not sure how intransparent this would be for new users, but wouldn't this be a prime case for StandardScaler? (And the function requires sklearn anyways)

File: examples/decoding/plot_ems_filtering.py (hunk 3 above)
Comment (jona-sassenhagen) [inline examples/decoding/plot_ems_filtering.py]: How about scaling it to STDs (Cohen's d)? Not 100% sure about STDs of *what* though.

File: mne/decoding/ems.py [hunk 5]
```diff
@@ -76,38 +179,36 @@ def compute_ems(epochs, conditions=None, picks=None, n_jobs=1, verbose=None):
...
+    # Setup cross-validation. Need to use _set_cv to deal with sklearn
+    # deprecation of cv objects.
+    y = epochs.events[:, 2]
+    _, cv_splits = _set_cv(cv, clf=LogisticRegression(), X=y, y=y)
 
     parallel, p_func, _ = parallel_func(_run_ems, n_jobs=n_jobs)
+    # FIXME this parallization should be removed.
+    #   1) it's numpy computation so it's already efficient,
+    #   2) it duplicates the data in RAM,
+    #   3) the computation is already super fast.
```
Comment (jona-sassenhagen) [inline mne/decoding/ems.py]: Agree. This is all trivial linalg stuff that takes no time. My hunch would be the parallelisation slows it down for most cases.

File: mne/decoding/tests/test_ems.py [hunk 6]
```diff
@@ -53,4 +54,34 @@ def test_ems():
...
+    epochs.equalize_event_counts(epochs.event_id)
+    if check_version('sklearn', '0.18'):
+        from sklearn.model_selection import StratifiedKFold
+        cv = StratifiedKFold()
+    else:
+        from sklearn.cross_validation import StratifiedKFold
+        cv = StratifiedKFold(len(epochs))
+    compute_ems(epochs, cv=cv)
+    compute_ems(epochs, cv=2)
+    compute_ems(epochs, cv=StratifiedKFold())
```
Comment (jona-sassenhagen) [inline mne/decoding/tests/test_ems.py]: Loop?

File: mne/decoding/tests/test_ems.py [hunk 7]
```diff
@@ -53,4 +54,34 @@ def test_ems():
...
+    assert_raises(ValueError, compute_ems, epochs, cv='foo')
+    assert_raises(ValueError, compute_ems, epochs, cv=len(epochs) + 1)
```
Comment (jona-sassenhagen) [inline mne/decoding/tests/test_ems.py]: *maybe* loop

Comment (kingjr) [inline mne/decoding/tests/test_ems.py]: no line gain..

Comment (jona-sassenhagen) [inline mne/decoding/tests/test_ems.py]: Yeah but it makes it immediately obvious that only 1 param is changing for each iteration. Not everything I say is about code golfing for fewer lines :)

Comment (jona-sassenhagen) [inline mne/decoding/tests/test_ems.py]: Not saying stuff like this NEEDS to be looped.

File: mne/decoding/ems.py [hunk 8]
```diff
@@ -76,38 +179,36 @@ def compute_ems(epochs, conditions=None, picks=None, n_jobs=1, verbose=None):
...
-        if hasattr(LeaveOneOut, 'split'):
-            cv = LeaveOneOut()
-            return cv.split(np.zeros((n, 1)))
-        else:
-            cv = LeaveOneOut(len(data))
-            return cv
+    # Setup cross-validation. Need to use _set_cv to deal with sklearn
+    # deprecation of cv objects.
+    y = epochs.events[:, 2]
+    _, cv_splits = _set_cv(cv, clf=LogisticRegression(), X=y, y=y)
```
Comment (jona-sassenhagen) [inline mne/decoding/ems.py]: Not sure I understand this ... is the LR really only here so _set_cv knows it's supposed to pick a classification split CV?

Seems a bit inelegant, though I can't think of something better.

File: mne/decoding/ems.py [hunk 9]
```diff
@@ -1,18 +1,110 @@
...
+        """Transform the data by the spatial filters.
+
+        Parameters
+        ----------
+        X : array, shape (n_epochs, n_channels, n_times)
+            The input data.
+
+        Returns
+        -------
+        X : array, shape (n_epochs, n_filter, n_times)
```
Comment (jona-sassenhagen) [inline mne/decoding/ems.py]: n_filter**s** ?

File: mne/decoding/ems.py [hunk 10]
```diff
@@ -1,18 +1,110 @@
...
+    ----------
+    epochs : instance of mne.Epochs
+        The epochs.
+    conditions : list of str | None
+        If a list of strings, strings must match the
+        epochs.event_id's key as well as the number of conditions supported
+        by the objective_function. If None keys in epochs.event_id are used.
+    picks : array-like of int | None
+        Channels to be included. If None only good data channels are used.
+        Defaults to None
```
Comment (jona-sassenhagen) [inline mne/decoding/ems.py]: Okay, now I'm really nitpicking ... missing period

File: examples/decoding/plot_ems_filtering.py (hunk 3 above)
Comment (kingjr) [inline examples/decoding/plot_ems_filtering.py]: yes, but this is up to the authors, we just implement...

File: examples/decoding/plot_ems_filtering.py (hunk 4 above)
Comment (kingjr) [inline examples/decoding/plot_ems_filtering.py]: Yes, although
*  they don't remove the mean... 
* it's not a feature scaling, it's an overall scaling

Overall I thought this notation would be more explicit like this.

File: mne/decoding/ems.py (hunk 8 above)
Comment (kingjr) [inline mne/decoding/ems.py]: Yes, I agree; I didn't find a nice quick way of doing it. Suggestions welcome!

File: mne/decoding/ems.py (hunk 10 above)
Comment (kingjr) [inline mne/decoding/ems.py]: I'm ok with nitpicking ;)

File: examples/decoding/plot_ems_filtering.py (hunk 4 above)
Comment (jona-sassenhagen) [inline examples/decoding/plot_ems_filtering.py]: Explicit yes, but less sklearn pipeline-y ...

`sklearn.preprocessing.StandardScaler(with_mean=False)`

I agree, I see pros and cons though.

Comment (kingjr) [inline examples/decoding/plot_ems_filtering.py]: I agree. I think StandardScaler is confusing here mainly because we would apply it on a X.ravel(), and not within a pipeline

File: mne/decoding/ems.py (hunk 5 above)
Comment (kingjr) [inline mne/decoding/ems.py]: We'd need to deprecate n_jobs, so for another PR IMO

File: mne/decoding/ems.py [hunk 11]
```diff
@@ -1,18 +1,110 @@
...
+
+        Returns
+        -------
+        X : array, shape (n_epochs, n_filter, n_times)
+            The input data transformed by the spatial filters.
+        """
+        Xt = np.sum(X * self.filters_, axis=1)
+        # Xt = [np.dot(filt.T, x.T) for filt, x in
+        #       zip(self.filters_.T, X.transpose(2, 0, 1))]
+        # Xt = np.transpose(Xt)
```
Comment (agramfort) [inline mne/decoding/ems.py]: cleanup?

**Comment by agramfort** [comment]
see what CIs are not happy.

besides looks good enough for me.

**Comment by kingjr** [comment]
@agramfort ready on my end

File: examples/decoding/plot_ems_filtering.py [hunk 12]
```diff
@@ -57,38 +60,71 @@
...
+plt.ylabel('a.u.')
 plt.legend(loc='best')
-
-
-# visualize spatial filters across time
 plt.show()
+
+# Visualize spatial filters across time
 evoked = epochs.average()
 evoked.data = filters
```
Comment (agramfort) [inline examples/decoding/plot_ems_filtering.py]: can you use EvokedArray rather than hacking the data attribute?

File: mne/decoding/ems.py [hunk 13]
```diff
@@ -1,26 +1,121 @@
...
 
+from .mixin import TransformerMixin, EstimatorMixin
+from .base import _set_cv
 from ..utils import logger, verbose
 from ..fixes import Counter
 from ..parallel import parallel_func
 from .. import pick_types, pick_info
 
 
+class EMS(TransformerMixin, EstimatorMixin):
```
Comment (agramfort) [inline mne/decoding/ems.py]: can you add a repr ?

In [3]: ems
Out[3]: <mne.decoding.ems.EMS at 0x119fc5810>

so far.

File: mne/decoding/ems.py [hunk 14]
```diff
@@ -1,26 +1,121 @@
...
+        epochs.event_id's key as well as the number of conditions supported
+        by the objective_function. If None keys in epochs.event_id are used.
+    picks : array-like of int | None
+        Channels to be included. If None only good data channels are used.
+        Defaults to None.
+    n_jobs : int
+        Number of jobs to run in parallel.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+        Defaults to self.verbose.
```
Comment (agramfort) [inline mne/decoding/ems.py]: what are these parameters? there is no init in this object

File: mne/decoding/ems.py [hunk 15]
```diff
@@ -1,26 +1,121 @@
...
+        Returns
+        -------
+        self : returns and instance of self.
+        """
+        classes = np.unique(y)
+        if len(classes) != 2:
+            raise ValueError('EMS only works for binary classification.')
+        self.classes_ = classes
+        filters = X[y == classes[0]].mean(0) - X[y == classes[1]].mean(0)
+        filters /= np.sqrt(np.sum(filters ** 2, axis=0))[None, :]
```
Comment (agramfort) [inline mne/decoding/ems.py]: use np.linalg.norm(..., axis=0)

File: mne/decoding/ems.py [hunk 16]
```diff
@@ -1,26 +1,121 @@
...
+
+        Parameters
+        ----------
+        X : array, shape (n_epochs, n_channels, n_times)
+            The input data.
+
+        Returns
+        -------
+        X : array, shape (n_epochs, n_filters, n_times)
+            The input data transformed by the spatial filters.
```
Comment (agramfort) [inline mne/decoding/ems.py]: I think the shape is wrong

File: mne/decoding/ems.py [hunk 17]
```diff
@@ -1,26 +1,121 @@
...
+        """
+        Xt = np.sum(X * self.filters_, axis=1)
+        return Xt
+
+
 @verbose
-def compute_ems(epochs, conditions=None, picks=None, n_jobs=1, verbose=None):
-    """Compute event-matched spatial filter on epochs
+def compute_ems(epochs, conditions=None, picks=None, n_jobs=1, verbose=None,
+                cv='LeaveOneOut'):
```
Comment (agramfort) [inline mne/decoding/ems.py]: I would put cv=None and say in docstring it defaults to LOO

having cv as a string is not standard

File: mne/decoding/ems.py (hunk 5 above)
Comment (agramfort) [inline mne/decoding/ems.py]: ok for you @dengemann ?

**Comment by agramfort** [comment]
that's it for me

File: mne/decoding/ems.py (hunk 14 above)
Comment (kingjr) [inline mne/decoding/ems.py]: woops indeed. No param needed.

File: mne/decoding/ems.py (hunk 16 above)
Comment (kingjr) [inline mne/decoding/ems.py]: yes

**Comment by kingjr** [comment]
@dengemann do you want to give it a look?

File: examples/decoding/plot_ems_filtering.py [hunk 18]
```diff
@@ -57,38 +60,70 @@
...
-
-# visualize spatial filters across time
 plt.show()
-evoked = epochs.average()
-evoked.data = filters
+
+# Visualize spatial filters across time
+evoked = EvokedArray(filters, epochs.info, tmin=epochs.tmin)
 evoked.plot_topomap(ch_type=ch_type)
+
```
Comment (agramfort) [inline examples/decoding/plot_ems_filtering.py]: insert here line like this:

#############################################################

so it renders on the doc and notebook output as a separate section and images are then plotted at the end of each section rather than at the top.

File: mne/decoding/ems.py [hunk 19]
```diff
@@ -1,26 +1,111 @@
...
+    ----------
+    filters_ : ndarray, shape (n_channels, n_times)
+        The set of spatial filters.
+    classes_ : ndarray, shape (n_classes,)
+        The target classes.
+    """
+
+    def __repr__(self):
+        if hasattr(self, 'filters_'):
+            return 'EMS: fitted with %i filters on %i classes.' % (
```
Comment (agramfort) [inline mne/decoding/ems.py]: Please encapsulate the repr with <>

so just:

return '<EMS: fitted with %i filters on %i classes>' % ...

**Comment by agramfort** [comment]
besides LGTM

**Comment by agramfort** [comment]
+1 for merge when Cis are happy

**Comment by kingjr** [comment]
Ready for merge, thanks for the reviews

**Comment by dengemann** [comment]
Taking a look and merging if all is fine.

File: mne/decoding/ems.py [hunk 20]
```diff
@@ -1,26 +1,111 @@
...
+        Returns
+        -------
+        self : returns and instance of self.
+        """
+        classes = np.unique(y)
+        if len(classes) != 2:
+            raise ValueError('EMS only works for binary classification.')
+        self.classes_ = classes
+        filters = X[y == classes[0]].mean(0) - X[y == classes[1]].mean(0)
+        filters /= np.linalg.norm(filters, axis=0)[None, :]
```
Comment (dengemann) [inline mne/decoding/ems.py]: I think we had a norm fix somewhere in fixes. Not sure if it is still needed though. cc @Eric89GXL

File: mne/decoding/ems.py [hunk 21]
```diff
@@ -1,26 +1,111 @@
...
     be specified. The result is a spatial filter at each time point and a
     corresponding time course. Intuitively, the result gives the similarity
     between the filter at each time point and the data vector (sensors) at
     that time point.
 
+    .. note : EMS only works for binary classification.
+    .. note : The present function applies a leave-one-out cross-validation,
+              following Schurger et al's paper. However, we recommend using
+              a Stratified K-Fold cross-validation, which is known to have
+              less biases. Because of the leave-one-out, thise function needs
```
Comment (dengemann) [inline mne/decoding/ems.py]: Maybe we can be more precise here. Less biases sounds somehow a bit sloppy. Maybe say that LOO tends to overfit and generate too much variance etc. Also text in example.

Comment (dengemann) [inline mne/decoding/ems.py]: Stratified -> stratified ?

File: mne/decoding/ems.py [hunk 22]
```diff
@@ -76,38 +165,36 @@ def compute_ems(epochs, conditions=None, picks=None, n_jobs=1, verbose=None):
...
-    # special care to avoid path dependent mappings and orders
+    # Special care to avoid path dependent mappings and orders
     conditions = list(sorted(conditions))
     cond_idx = [np.where(ev == epochs.event_id[k])[0] for k in conditions]
 
     info = pick_info(epochs.info, picks)
     data = epochs.get_data()[:, picks]
 
     # Scale (z-score) the data by channel type
+    # XXX the z-scoring is applied outside the CV, which is not standard.
```
Comment (dengemann) [inline mne/decoding/ems.py]: should maybe be an option or so ... sounds more like a bug actually.

File: mne/decoding/ems.py (hunk 5 above)
Comment (dengemann) [inline mne/decoding/ems.py]: yep, let's do it.

**Comment by dengemann** [comment]
@kingjr @agramfort one minor concern is the example, it now gets more complex and longer. Any ideas how it could be simplified?

File: mne/decoding/ems.py (hunk 22 above)
Comment (kingjr) [inline mne/decoding/ems.py]: it's not a bug, it's the way they do it in the paper

**Comment by kingjr** [comment]
It's more complicated because it's less blackboxy. The last few lines are sufficient to reproduce the whole script.

File: mne/decoding/ems.py (hunk 22 above)
Comment (dengemann) [inline mne/decoding/ems.py]: I know...

**Comment by dengemann** [comment]
> It's more complicated because it's less blackboxy. The last few lines are sufficient to reproduce the whole script.

We recently said while restructuring the docs that we should have examples as short and simple as possible and only do such things in narrative tutorials. Then I'm not sure we want a tutorial for this. 
Just saying as an approximate value for orientation.

**Comment by kingjr** [comment]
@dengemann Ok, whatever you guys prefer. I just don't want to spend time on examples ATM, we need to move forwards in terms of enhancements.

File: mne/decoding/ems.py (hunk 22 above)
Comment (kingjr) [inline mne/decoding/ems.py]: ok, to be addressed in a different PR. This is just to make a sklearn compatible transformer

**Comment by agramfort** [comment]
I am fine with a bit longer examples if we use code blocks with sphinx-gallery

see how it renders now:

https://1906-1301584-gh.circle-artifacts.com/0//home/ubuntu/mne-python/doc/_build/html/auto_examples/decoding/plot_ems_filtering.html#sphx-glr-auto-examples-decoding-plot-ems-filtering-py
